nixos/tests/verity-protected-root.nix

```diff
@@ -0,0 +1,226 @@
+# Tests a NixOS system with a read-only root filesystem that's integrity-protected
+# through DM-verity. The root filesystem is mounted read-only, and for NixOS activation
+# to suceed, an overlay tmpfs is mounted on top of it.
+# This test uses systemd-repart to create a bootable disk image, as it supplies handy
+# utilities for creating verity partitions, but it can also be setup manually through
+# `systemd-veritysetup`.
+
+{ lib, pkgs, ... }:
+
+let
+  rootPartitionLabel = "root";
+  imageId = "verity-root-image";
+  imageVersion = "1-rc1";
+
+  # Use a random, but fixed root hash placeholder to allow us specifying the "real" root hash
+  # after the image is first built.
+  roothashPlaceholder = "61fe0f0c98eff2a595dd2f63a5e481a0a25387261fa9e34c37e3a4910edf32b8";
+in
+{
+  name = "verity-root";
+
+  meta.maintainers = with lib.maintainers; [ msanft ];
+
+  nodes.machine = { config, lib, pkgs, ... }: {
+
+    imports = [ ../modules/image/repart.nix ];
+
+    virtualisation.directBoot.enable = false;
+    virtualisation.mountHostNixStore = false;
+    virtualisation.useEFIBoot = true;
+
+    # Disable boot loaders, as a UKI is used, which contains systemd-stub.
+    # TODO(raitobezarius): revisit this when #244907 lands
+    boot.loader.grub.enable = false;
+
+    system.image.id = imageId;
+    system.image.version = imageVersion;
+
+    # Override defaults from the test harness.
+    # Normally, users won't need to do this, as the option defaults to an empty attrset.
+    fileSystems = lib.mkForce { };
+    virtualisation.fileSystems = lib.mkForce { };
+
+    # Provides 'veritysetup' to check if the verity-protected device
+    # has been mapped correctly.
+    environment.systemPackages = with pkgs; [ cryptsetup ];
+
+    boot.initrd = {
+      kernelModules = [ "overlay" ];
+      supportedFilesystems = [ "erofs" ];
+
+      # This adds some udev rules that are required for dm-verity to work.
+      services.lvm.enable = true;
```

Comment:
I find it odd that you need to enable this yourself. This should be enabled when you enable dm-verity support.

Reply:
Yeah, I agree. But looking around, I think it's uncommon to have modules enable things in other modules. Is there some kind of agreement around this?